### PR TITLE
modern-cpp-kafka: new recipe

### PR DESCRIPTION
**Pull request by kenneth-jia**
Specify library name and version:  **modern-cpp-kafka/v2022.06.15**

Hi,
I'm the author of `modern-cpp-kafka` (https://github.com/morganstanley/modern-cpp-kafka), a Kafka C++ API, wrapper over `librdkafka`).
Now I'm trying to add this header-only library to the ConanCenter (https://github.com/morganstanley/modern-cpp-kafka/issues/105)
Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by toge** [comment]
@kenneth-jia 
Hello.
Thanks for creating the conan recipe!

Unfortunately, the recipe you wrote must have test_package folder.
There is a recently created template as a sample of the recipe, could you please refer this as a reference to modify it?
https://github.com/conan-io/conan-center-index/tree/master/docs/package

**Comment by kenneth-jia** [comment]
> @kenneth-jia Hello. Thanks for creating the conan recipe!
> 
> Unfortunately, the recipe you wrote must have test_package folder. There is a recently created template as a sample of the recipe, could you please refer this as a reference to modify it? https://github.com/conan-io/conan-center-index/tree/master/docs/package

Just added the `test_package` part, and now is pending on the access approval. Thanks!

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/conanfile.py:14]:
```suggestion
    settings = "arch", "build_type", "compiler", "os"
```
There are not header to export here. Please, read https://docs.conan.io/en/latest/creating_packages/package_repo.html#exporting-the-sources-with-the-recipe-exports-sources to understand exports_sources usage.

Keep full settings always. You can customize the package ID on `package_id` method.

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/conanfile.py:30]: Package info is missing. Conan need to generate `RdKafka::rdkafka` according.

Please, read https://docs.conan.io/en/latest/creating_packages/package_information.html#using-components and https://docs.conan.io/en/latest/migrating_to_2.0/properties.html

Also, pthread is a mandatory dependency https://github.com/morganstanley/modern-cpp-kafka/blob/main/CMakeLists.txt#L50

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/test_package/conanfile.py:8]: Why do you need only for testing and it's not a dependency for modern-cpp-kafka?

File: recipes/modern-cpp-kafka/config.yml
Comment (uilianries) [inline recipes/modern-cpp-kafka/config.yml:2]:
```suggestion
  "2022.06.15":
```

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/conanfile.py:11]:
```suggestion
    topics = ("kafka", "librdkafka", "kafkaproducer", "kafkaconsumer")
```
conan and the project name are not valid topics. Topics are keywords when searching a related feature

**Review by uilianries — CHANGES_REQUESTED** [review]
Thank you for your contribution

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/test_package/conanfile.py:8]: Oh, that's the tricky part -- we don't want to bind the `modern-cpp-kafka` (header-only API) to a specific `librdkafka` (the underlying C library) version. E.g. `modern-cpp-kafka/2022.06.15` could work together with `librdkafka/1.7.0`, `librdkafka/1.8.0`, `librdkafka/1.8.2` and `librdkafka/1.9.0 (not available in conan center yet)` , and user might want to pickup just a specific `librdkafka` version (not the one which `modern-cpp-kafka` depends on), -- some `librdkafka` version might have a few defects and users don't always want to pick up the latest one.

Then, the `connanfile.py` (for user's application) might be,
```
def requirements(self):
    self.requires("librdkafka/1.7.0")
    self.requires("modern-cpp-kakfa/2022.06.15")
```

File: recipes/modern-cpp-kafka/config.yml
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/config.yml]: Fixed in the latest commit.

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/conanfile.py]: Fixed in the latest commit.

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/conanfile.py]: Fixed in the latest commit.

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/conanfile.py:30]: Hi, @uilianries I just updated the commit. Please help review again. Thank you a lot! 😀

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (uilianries) [reply recipes/modern-cpp-kafka/all/test_package/conanfile.py:8]: Okay, it sounds fine. We have other few projects which follow the same line.

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/conanfile.py:26]:
```suggestion
        self.copy(pattern="*.h", dst="include", src=os.path.join(self._source_subfolder, "include"))
```
filter only headers. There is a cmake file the middle

File: recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt:7]:
```suggestion
find_package(RdKafka REQUIRED CONFIG)
find_package(ModernCppKafka REQUIRED CONFIG)
```

File: recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt:10]:
```suggestion
target_link_libraries(${PROJECT_NAME} RdKafka::rdkafka ModernCppKafka::ModernCppKafka)
```

File: recipes/modern-cpp-kafka/all/conanfile.py
Comment (uilianries) [inline recipes/modern-cpp-kafka/all/conanfile.py:10]:
```suggestion
    license = "Apache-2.0"
```
Use short SPDX always: https://spdx.org/licenses/Apache-2.0.html

**Review by uilianries — COMMENTED** [review]
only a small detail

**Review by uilianries — APPROVED** [review]
LGTM

Just need to wait for user approval.

**Comment by kenneth-jia** [comment]
> LGTM
> 
> Just need to wait for user approval.

Thank you a lot for the help, @uilianries! 😀

**Comment by uilianries** [comment]
@kenneth-jia No, thank you for contributing! 💚

**Comment by kenneth-jia** [comment]
Hi, @uilianries 
I just updated the PR with a small commit for the test_package build (https://github.com/conan-io/conan-center-index/pull/11430/commits/3fc99027fc557dcfcf03ab5be763c6d41e10657c). Could you please help review again? Thanks! :-)

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/conanfile.py:9]: If `modern-cpp-kafka` requires `librdkafka`, it should be a requirement of the recipe itself (yes, even if it is a header only package), that way the user consuming `modern-cpp-kafka` has everything needed (see `approvaltests.cpp`, it is the same pattern).

If the consumer wants to override the `librdkafka` they can always do it explicitly:

```python
def requirements(self):
    self.requires("librdkafka/<user-overrides>")
    self.requires("modern-cpp-kakfa/2022.06.15")
```

but if they are fine with the hardcoded version, they can just require what they directly need:

```python
def requirements(self):
    self.requires("modern-cpp-kakfa/2022.06.15")
```

**Review by jgsogo — CHANGES_REQUESTED** [review]

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/test_package/conanfile.py:9]: @jgsogo
OH, users are free to override the dependencies explicitly? That's flexible for conan! Let me update the PR. Thanks!

File: recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt:13]:
```suggestion
target_link_libraries(${PROJECT_NAME} ModernCppKafka::ModernCppKafka)
```

This should arrive as a transitive dependency

File: recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt:9]:
```suggestion
```

This is a transitive dep

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/conanfile.py:14]: Let me know if the message from the crawler is good enough to fix the error shown here.

File: recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/CMakeLists.txt:7]: If the `cpp` wrapper requires 17, we need a `validate()` method in the recipe to check the compiler supports this standard.

**Review by jgsogo — COMMENTED** [review]
This is looking good

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/test_package/conanfile.py]: I have to use `# pylint: skip-file` for the v1 test, otherwise it shows `tools.cross_building` could not be found.

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/conanfile.py:1]:
```suggestion
```

This is only allowed for v1-only tests, which should use `test_v1_****` pattern. Below I suggest the [required changes](https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md#import-tools-from-conan)

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/conanfile.py:2]:
```suggestion
from conans import ConanFile, CMake
from conan.tools.build import cross_building
```

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (jgsogo) [inline recipes/modern-cpp-kafka/all/test_package/conanfile.py:15]:
```suggestion
        if not cross_building(self):
```

**Review by jgsogo — CHANGES_REQUESTED** [review]

File: recipes/modern-cpp-kafka/all/test_package/conanfile.py
Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/test_package/conanfile.py:1]: I'm wondering whether there's anything missed, -- if I changed the `test_package` folder to `test_v1_package` (together with the `bin_path` for the test), the test package would not be found! 🤐

Comment (kenneth-jia) [reply recipes/modern-cpp-kafka/all/test_package/conanfile.py:1]: > I'm wondering whether there's anything missed, -- if I changed the `test_package` folder to `test_v1_package` (together with the `bin_path` for the test), the test package would not be found! 🤐

Oh, I see, -- since it's not `ready for Conan v2` (I tried to migrate, but failed and have to give up), we don't need the `pylint` line either. Let me have a try. Thanks!😀

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**
